symengine/series.cpp

```diff
@@ -0,0 +1,173 @@
+#include <exception>
+#include <algorithm>
+#include <iterator>
+#include <symengine/series.h>
+#include <symengine/dict.h>
+
+using SymEngine::RCP;
+using SymEngine::make_rcp;
+
+namespace SymEngine {
+
+UnivariateSeries::UnivariateSeries(const RCP<const Symbol> &var, const unsigned int &precision, const RCP<const UnivariatePolynomial> &poly) :
+        var_{var}, poly_{std::move(poly)} , prec_{precision} {
+    SYMENGINE_ASSERT(UnivariateSeries::is_canonical(*poly_, prec_))
+}
+
+UnivariateSeries::UnivariateSeries(const RCP<const Symbol> &var, const unsigned int &precision, const unsigned int &max, map_uint_mpz&& dict) :
+        var_{var}, prec_{precision} {
+
+    poly_ = univariate_polynomial(var_, max, std::move(dict));
+
+    SYMENGINE_ASSERT(UnivariateSeries::is_canonical(*poly_, prec_))
+}
+
+UnivariateSeries::UnivariateSeries(const RCP<const Symbol> &var, const unsigned int &precision, const map_uint_mpz& dict) :
+        var_{var}, prec_{precision} {
+
+    map_uint_mpz dict_trunc;
+    unsigned int max = 0;
+    std::copy_if(dict.begin(), dict.end(), std::inserter(dict_trunc, dict_trunc.end()),
+        [&](const map_uint_mpz::value_type i)
+        {
+            if (i.first < prec_) {
+                if (max < i.first)
+                    max = i.first;
+                return true;
+            }
+            return false;
+        } );
+    poly_ = univariate_polynomial(var_, max, std::move(dict_trunc));
+
+    SYMENGINE_ASSERT(UnivariateSeries::is_canonical(*poly_, prec_))
+}
+
+UnivariateSeries::UnivariateSeries(const RCP<const Symbol> &var, const unsigned int &precision, const std::vector<mpz_class> &v) :
+        var_{var}, prec_{precision} {
+
+    std::vector<mpz_class> vtrunc;
+    std::copy_if(v.begin(), v.end(), std::back_inserter(vtrunc),
+        [&](decltype(v[0]) i) { return i < prec_; } );
+    poly_ = UnivariatePolynomial::create(var_, vtrunc);
+
+    SYMENGINE_ASSERT(UnivariateSeries::is_canonical(*poly_, prec_))
+}
+
+bool UnivariateSeries::is_canonical(const UnivariatePolynomial& poly, const unsigned int &prec)
+{
+    return true;
```

Comment:
Given the internal representation of the series, can it ever be not canonical? If not, then I would completely remove this `is_canonical` function and tests above, since they don't test anything.